### PR TITLE
Search content items by title and base_path

### PR DESCRIPTION
**Pull request by dougdroper**
If searching for a field that is not supported, that field is ignored
and it doesn't change the behaviour of the rest of the query

https://trello.com/c/9rpo1V2u/683-search-functionality-on-get-content-collection-endpoint-searching-documents-by-slug-and-title-in-sp

### CODE REVIEW

**Comment by danielroseman** [comment]
I don't think this quite supports the use case of specialist-publisher, which is to search for a single string and match against both title and slug.

**Comment by dougdroper** [comment]
@danielroseman ah I see, I thought they needed some level of granularity. In the case for a single string, they could just call the API with q_title=SEARCH&q_base_path=SEARCH which will match SEARCH in title AND base_path, or do they need an interface that just does q=SEARCH and match against title OR base_path?

**Comment by tuzz** [comment]
If an item previously had a title of `foo` and it changed to `bar` and was published, what would happen if you search for `q_title=foo`? Would it be an empty set of results or would it include the superseded item?

**Comment by danielroseman** [comment]
@dougdroper the latter, it needs to be OR.

**Comment by dougdroper** [comment]
@tuzz If the latest content_item has a title of `foo`, then searching for `bar` would return an empty set. This is just reducing the scope of the existing query, ie, it still gets passed to `GetLatest`

**Comment by dougdroper** [comment]
@danielroseman updated, let me know if that supports the specialist-publishers needs.

File: spec/controllers/v2/content_items_controller_spec.rb [hunk 1]
```diff
@@ -45,6 +45,27 @@
       )
     end
 
+    context "searching a field" do
+      context "when there is a valid query" do
+        let!(:content_item) do
+          FactoryGirl.create(:live_content_item,
+                             base_path: "/foo",
+                             format: "topic",
+                             title: "foo bar")
```
Comment (danielroseman) [inline spec/controllers/v2/content_items_controller_spec.rb]: Would be better remove "foo" from the title so that there is no overlap between the fields.

**Comment by danielroseman** [comment]
Looks good, one minor comment and there is a linting issue.

**Comment by tuzz** [comment]
That's not quite what I asked. I'm not sure what will happen if you search for a title that is no longer the title of the latest content item. Could we possibly write a test to check this?

I'm not sure what the behaviour of this should be, either. Whether we should return the superseded item or whether it should be scoped to only the latest (draft/published) versions.

File: app/presenters/queries/content_item_presenter.rb [hunk 2]
```diff
@@ -34,22 +34,31 @@ def initialize(scope, params = {})
...
       end
 
       def present_many
-        scope = ::Queries::GetLatest.call(self.scope)
-        scope = join_supporting_objects(scope)
-        scope = order_and_paginate(scope)
-
-        extract_fields(scope)
+        parsing_enumerator(evaluate_query(ordered_extracted_fields))
       end
```
Comment (tuzz) [inline app/presenters/queries/content_item_presenter.rb]: I find it quite difficult to read these nested method calls. I found the procedural style easier to read, but maybe that's just me? By breaking everything into one line methods, I personally struggle to build a global understanding of what this class does.

Comment (dougdroper) [inline app/presenters/queries/content_item_presenter.rb]: Yes I suppose this particular method is a little tricky to parse inside out. I was trying to expose the ordering of these methods, but have done a bit more refactoring in f997d4297f23157e18fcab20c13d6a1f827b1a4e to try and express the intent more clearly. 

I guess from a global understanding of the class I was trying to sum up what this does in the one public method, and then encapsulate the complexity lower down. When I read this class I want to understand what it does as fast as possible without looking below `private` so tend to prefer this style, but it certainly increases the amount you have to jump around, but hopefully with clearer method names it is clearer overall. Lets go through this class together and see if there is a way to make it clearer.

Comment (tuzz) [inline app/presenters/queries/content_item_presenter.rb]: Ok, sure. Let's go through this when you're at a natural breaking point. 😃

**Comment by tuzz** [comment]
Docs for that endpoint are here:
https://github.com/alphagov/publishing-api/blob/master/doc/publishing-api-syntactic-usage.md#get-v2content

We might also want to consider adding a pact to gds-api-adapters:
https://github.com/alphagov/gds-api-adapters/blob/master/test/publishing_api_v2_test.rb#L906

File: app/presenters/queries/content_item_presenter.rb [hunk 3]
```diff
@@ -86,10 +100,11 @@ def extract_fields(scope)
...
 
-        scope = scope.select(*fields_to_select)
-        results = evaluate_query(scope)
+        scope.select(*fields_to_select)
+      end
 
-        parsing_enumerator(results)
+      def search(scope)
+        scope.where("title ilike ? OR base_path ilike ?", "%#{search_query}%", "%#{search_query}%")
       end
```
Comment (tuzz) [inline app/presenters/queries/content_item_presenter.rb]: What happens if no search query is provided? Will it still attach this scope and does that impact performance?

Comment (dougdroper) [inline app/presenters/queries/content_item_presenter.rb]: Good catch 👍  my guard clause seems to have disappeared in the refactoring. Back now.

**Pull request merged**